File: crates/openduo-tools/src/labels.rs
use crate::registry::Tool;
use anyhow::Result;
use async_trait::async_trait;
use openduo_core::gitlab_client::GitLabClient;
use serde_json::{json, Value};

pub struct LabelTools;

impl LabelTools {
    pub fn all(client: GitLabClient) -> Vec<Box<dyn Tool>> {
        vec![
            Box::new(ListLabels { client: client.clone() }),
            Box::new(CreateLabel { client: client.clone() }),
        ]
    }
}

struct ListLabels { client: GitLabClient }
#[async_trait]
impl Tool for ListLabels {
    fn name(&self) -> &str { "list_labels" }
    fn description(&self) -> &str { "List labels for a GitLab project." }
    fn parameters_schema(&self) -> Value { json!({ "type": "object", "properties": { "project_id": { "type": "string" } }, "required": ["project_id"] }) }
    async fn execute(&self, args: Value) -> Result<String> {
        let pid = urlencoding::encode(args["project_id"].as_str().unwrap_or_default());
        let v: Vec<Value> = self.client.get(&format!("projects/{}/labels", pid)).await?;
        Ok(serde_json::to_string_pretty(&v)?)
    }
}

struct CreateLabel { client: GitLabClient }
#[async_trait]
impl Tool for CreateLabel {
    fn name(&self) -> &str { "create_label" }
    fn description(&self) -> &str { "Create a new label in a GitLab project." }
    fn parameters_schema(&self) -> Value { json!({ "type": "object", "properties": { "project_id": { "type": "string" }, "name": { "type": "string" }, "color": { "type": "string", "description": "Hex color code e.g. #FF0000" } }, "required": ["project_id", "name", "color"] }) }
    async fn execute(&self, args: Value) -> Result<String> {
        let pid = urlencoding::encode(args["project_id"].as_str().unwrap_or_default());
        let v: Value = self.client.post(&format!("projects/{}/labels", pid), args).await?;
        Ok(serde_json::to_string_pretty(&v)?)
    }
}

File: extension/server/tools/labels.ts
import type { GitLabClient } from '../gitlabClient';
import type { Tool } from './tool';

function enc(s: string): string {
  return encodeURIComponent(s);
}

export function labelTools(client: GitLabClient): Tool[] {
  return [
    {
      name: 'list_labels',
      description: 'List labels for a GitLab project.',
      parametersSchema: () => ({
        type: 'object',
        properties: {
          project_id: { type: 'string' },
        },
        required: ['project_id'],
      }),
      async execute(args) {
        const pid = enc(args.project_id as string);
        return JSON.stringify(await client.get(`projects/${pid}/labels`), null, 2);
      },
    },
    {
      name: 'create_label',
      description: 'Create a new label in a GitLab project.',
      parametersSchema: () => ({
        type: 'object',
        properties: {
          project_id: { type: 'string' },
          name: { type: 'string' },
          color: { type: 'string', description: 'Hex color code e.g. #FF0000' },
        },
        required: ['project_id', 'name', 'color'],
      }),
      async execute(args) {
        const pid = enc(args.project_id as string);
        const body: Record<string, unknown> = {};
        if (args.name) body.name = args.name;
        if (args.color) body.color = args.color;
        return JSON.stringify(await client.post(`projects/${pid}/labels`, body), null, 2);
      },
    },
  ];
}
